lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -0,0 +1,783 @@
+/*
+ * we assume good default setting of write concern is good for all
+ * bulk writes. Note that bulk writes are not transactions but ordered
+ * writes. They may fail in between. To some extend those situations
+ * may generate orphans but not alter the proper conduct of operations
+ * (what he user wants and what we acknowledge to the user).
+ *
+ * Orphan situations may be recovered by the Lifecycle.
+ *
+ * We use proper atomic operations when needed.
+ */
+const async = require('async');
+
+const constants = require('../../../constants');
+
+const errors = require('../../../errors');
+const BucketInfo = require('../../../models/BucketInfo');
+
+const MongoClient = require('mongodb').MongoClient;
+const Uuid = require('uuid');
+const diskusage = require('diskusage');
+
+const genVID = require('../../../versioning/VersionID').generateVersionId;
+const listAlgos = require('../../../algos/list/exportAlgos');
+
+const MongoReadStream = require('./readStream');
+const MongoUtils = require('./utils');
+
+const USERSBUCKET = '__usersbucket';
+const METASTORE = '__metastore';
+const INFOSTORE = '__infostore';
+const __UUID = 'uuid';
+const ASYNC_REPAIR_TIMEOUT = 15000;
+
+let uidCounter = 0;
+
+const VID_SEP = require('../../../versioning/constants')
+    .VersioningConstants.VersionId.Separator;
+
+function generateVersionId(replicationGroupId) {
+    // generate a unique number for each member of the nodejs cluster
+    return genVID(`${process.pid}.${uidCounter++}`,
+                  replicationGroupId);
+}
+
+function formatVersionKey(key, versionId) {
+    return `${key}${VID_SEP}${versionId}`;
+}
+
+function inc(str) {
+    return str ? (str.slice(0, str.length - 1) +
+            String.fromCharCode(str.charCodeAt(str.length - 1) + 1)) : str;
+}
+
+const VID_SEPPLUS = inc(VID_SEP);
+
+function generatePHDVersion(versionId) {
+    return {
+        isPHD: true,
+        versionId,
+    };
+}
+
+
+/**
+ * @constructor
+ *
+ * @param {Object} params - constructor params
+ * @param {String} params.host - host for mongo
+ * @param {String} params.port - port for mongo
+ * @param {String} params.replicationGroupId - replication group id
+ * used here to generate version id's
+ * // Does backbeat use this at all? Can we make this optional and
```

Comment:
@jonathan-gramain or @rahulreddy, do you use the concept of replication group id?